// eslint-plugin-flow-check/library/flow.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _child_process = require('child_process');

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _flowBin = require('flow-bin');

var _flowBin2 = _interopRequireDefault(_flowBin);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _toArray(arr) { return Array.isArray(arr) ? arr : Array.from(arr); }

exports.default = flow;


(0, _child_process.spawn)(_flowBin2.default, ['start']); // start a flow server

function flow(options) {
	var result = (0, _child_process.spawnSync)(_flowBin2.default, ['check-contents', '--json', '--root=' + options.root, options.fileName], {
		input: options.source,
		encoding: 'utf-8'
	});

	if (result.status !== 0) {
		return [{
			message: result.stderr,
			loc: {
				start: {
					line: 1
				},
				end: {
					line: 1
				}
			}
		}];
	}

	var _JSON$parse = JSON.parse(result.stdout),
	    errors = _JSON$parse.errors;

	return errors.map(function (error) {
		var _error$message = _toArray(error.message),
		    leading = _error$message[0],
		    rest = _error$message.slice(1);

		var payload = rest.map(function (m) {
			return format(m, {
				root: options.root
			});
		});

		return {
			message: [leading.descr].concat(_toConsumableArray(payload)).join(': '),
			path: leading.path,
			start: leading.loc.start.line,
			end: leading.loc.end.line,
			loc: leading.loc
		};
	});
}

function format(message, options) {
	if (message.type === 'Comment') {
		return message.descr;
	}
	if (message.type === 'Blame') {
		var address = message.path ? _path2.default.relative(options.root, message.path) + ':' + message.line : message.line;

		var see = message.path ? 'See ' + address : 'See line ' + address;
		return ['\'' + message.descr + '\'', see].join('. ');
	}
	return '\'' + message.descr + '\'';
}
module.exports = exports['default'];
